src/StrokeEngine.cpp

```diff
@@ -1,551 +1,177 @@
+      break;
+  }
+  
+  this->sendParameters(this->_patternIndex);
+  
+  ESP_LOGD("StrokeEngine", "Stroke Parameter %s - %f", name, debugValue);
+  
+  // When running a pattern and immediate update requested: 
+  if (applyNow == true) {
+    this->applyUpdate = true;
+    ESP_LOGD("StrokeEngine", "Setting Apply Update Flag!");
+  }
+
+  xSemaphoreGive(_patternMutex);
+}
+
+// TODO - Better naming?
+void StrokeEngine::sendParameters(int patternIndex) {
+  // Stroke is dynamically constrainted based on depth.
+  // This allows depth to change, and allow stroke to fill the available space, 
+  // rather than having to set the parameter again
+
+  patternTable[patternIndex]->setSpeedLimit(
+    this->motor->getMaxSpeed(), 
+    this->motor->getMaxAcceleration(),
+    1
+  );
+
+  // TODO - Handle RATE and SPEED here
+  patternTable[patternIndex]->setTimeOfStroke(constrain(60.0 / this->strokeRate, 0.01, 120.0));
+  patternTable[patternIndex]->setStroke(constrain(int(this->stroke), 0, this->depth));
+  patternTable[patternIndex]->setDepth(this->depth);
+  patternTable[patternIndex]->setSensation(this->sensation);
+}
+
+float StrokeEngine::getParameter(StrokeParameter parameter) {
+  switch (parameter) {
+    // TODO - Add SPEED
+    case StrokeParameter::RATE:
+      return this->strokeRate;
+    case StrokeParameter::DEPTH:
+      return this->depth;
+    case StrokeParameter::STROKE:
+      return this->stroke;
+    case StrokeParameter::SENSATION:
+      return this->sensation;
+    case StrokeParameter::PATTERN:
+      return this->_patternIndex;
+  }
 }
 
-void StrokeEngine::enableAndHome(endstopProperties *endstop, float speed) {
-    // set homing pin as input
-    _homeingPin = endstop->endstopPin;
-    pinMode(_homeingPin, endstop->pinMode);
-    _homeingActiveLow = endstop->activeLow;
-    _homeingSpeed = speed * _motor->stepsPerMillimeter;
-
-    // set homing direction so sign can be multiplied
-    if (endstop->homeToBack == true) {
-        _homeingToBack = 1;
-    } else {
-        _homeingToBack = -1;
-    }
-
-    // first stop current motion and delete stroke task
-    stopMotion();
-
-    // Enable Servo
-    servo->enableOutputs();
-
-    // Create homing task
+bool StrokeEngine::startPattern() {
+  // Only valid if state is ready
+  if (!this->motor->isInState(MotorState::ACTIVE)) {
+    ESP_LOGE("StrokeEngine", "Failed to start pattern! Motor is not active!");
+    return false;
+  }
+
+  Pattern* pattern = patternTable[_patternIndex];
+  ESP_LOGE("StrokeEngine", "Starting pattern %s", pattern->getName());
+
+  // Stop current move, should one be pending (moveToMax or moveToMin)
+  if (this->motor->hasStatusFlag(MOTOR_FLAG_MOTION_ACTIVE)) {
+    this->motor->stopMotion();
+  }
+
+  // Reset Stroke and Motion parameters
+  _index = -1;
+  if (xSemaphoreTake(_patternMutex, portMAX_DELAY) == pdTRUE) {
+    this->sendParameters(this->_patternIndex);        
+    xSemaphoreGive(_patternMutex);
+  }
+
+  #ifdef DEBUG_TALKATIVE

```

Comment:
You're using the esp_log. Should we unify this throughout the code?

Reply:
Yeah we definitely should. I wanna build a small `logging.h` which exposes some generic log handlers. I think OSSM does something similar. I just haven't gotten around to it yet